src/gui/TrackContentObjectView.cpp

```diff
@@ -1313,6 +1380,50 @@ TimePos TrackContentObjectView::draggedTCOPos( QMouseEvent * me )
 }
 
 
+int TrackContentObjectView::knifeMarkerPos( QMouseEvent * me )
+{
+	//Position relative to start of clip
+	const int markerPos = me->pos().x();
+
+	//In unquantized mode, we don't have to mess with the position at all
+	if ( unquantizedModHeld(me) ) { return markerPos; }
+	else
+	{	//Otherwise we...
+		//1: Convert the position to a TimePos
+		const float ppb = m_trackView->trackContainerView()->pixelsPerBar();
+		TimePos midiPos = markerPos * TimePos::ticksPerBar() / ppb;
+		//2: Snap to the correct position, based on modifier keys
+		midiPos = quantizeMarkerPos( midiPos, me->modifiers() & Qt::ShiftModifier );
+		//3: Convert back to a pixel position
+		return midiPos * ppb / TimePos::ticksPerBar();
+	}
+}
+
+
+
+
+TimePos TrackContentObjectView::quantizeMarkerPos( TimePos midiPos, bool shiftMode )
+{
+	const float snapSize = gui->songEditor()->m_editor->getSnapSize();
+	if ( shiftMode )
+	{	//If shift is held we quantize the length of the new left clip...
+		const TimePos leftPos = midiPos.quantize( snapSize );
+		//...or right clip...
+		const TimePos rightOff = m_tco->length() - midiPos;
+		const TimePos rightPos = m_tco->length() - rightOff.quantize( snapSize );
+		//...whichever gives a position closer to the cursor
+		if ( abs(leftPos - midiPos) < abs(rightPos - midiPos) ) { return leftPos; }
+		else { return rightPos; }
+	}
+	else
+	{
+		return TimePos(midiPos + m_initialTCOPos).quantize( snapSize ) - m_initialTCOPos;
```

Comment:
Why is it necessary to add `m_initialTCOPos` here and subtract it later? It seems it doesn't get added when `shiftMode` is true but when testing it the quantization worked on that mode too.

Reply:
Did you test with an offset clip? `midiPos` is relative to the clip's start position, so `m_midiPos + m_initialTCOPos` gives the absolute position of the cursor in the song. In other words:
- Shift mode quantizes cut position to a grid based on the start and end positions of the clip, choosing whichever is closest to the mouse. This  results in a left and/or right clip with a quantized length.
- Unmodified mode quantizes the cut position to the underlying song grid.

With apologies for the terrible colors, this image shows the unmodified snap positions in red, and shift snap in blue (start point based) and green (end point based)
![image](https://user-images.githubusercontent.com/8918426/103914775-a1bd6980-510a-11eb-9fbf-3157a3da9ac0.png)

Reply:
Oh, makes sense, thanks for explaining!